### PR TITLE
Stop computing outdated diagnostics with CancellationToken

### LINKED ISSUE #1263: Send Cancellation Token to TypeScript Language Server
[Cancellation token](https://code.visualstudio.com/api/references/vscode-api#cancellation-tokens) is issued when the next operation happens. For example, when typing `abc`, Vetur will query TS Server for diagnostics on `a`, `ab`, `abc`. If typing of `abc` finishes before `a`'s diagnostic request is responded, vscode-languageclient should send a cancellation token.

In this case, Vetur should forward that cancellation token to TS Server so it can stop computing outdated diagnostics/auto-completion. This should reduce a lot of CPU/memory usage from VLS.

### PR DESCRIPTION
**Pull request by yoyo930021**
Fixed #1263 
Fix a part of #2192

### CODE REVIEW

**Comment by yoyo930021** [comment]
I think we can't forward that cancellation token to TS Server for completion.
I only find a cancelling TS server all response method.

File: server/src/modes/script/javascript.ts [hunk 1]
```diff
@@ -107,18 +108,38 @@ export async function getJavascriptMode(
...
       ];
 
+      const compilerOptions = program.getCompilerOptions();
+      if (!!(compilerOptions.declaration || compilerOptions.composite)) {
+        rawScriptDiagnostics = [
+          ...rawScriptDiagnostics,
+          ...program.getDeclarationDiagnostics(sourceFile, cancellationToken?.tsToken)
+        ];
+      }
+
```
Comment (yoyo930021) [inline server/src/modes/script/javascript.ts:142]: Ref: https://github.com/microsoft/TypeScript/blob/88c24e6b0d0efb868088ed6da359ef55e6f1c6b6/src/services/services.ts#L1504

**Comment by yoyo930021** [comment]
When this PR merged:
- Replace `CLIEngine` to `Eslint` for eslint deprecated Node api.
- Add `getSuggestionDiagnostics` support. Like deprecated, unused.
- Add Fix all feature.

**Comment by yoyo930021** [comment]
This PR is broken VTI diagnostic.
I need more research.

**Comment by yoyo930021** [comment]
I made a retarded mistake, and I've fixed it.

File: server/src/utils/cancellationToken.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+export class VCancellationTokenSource extends CancellationTokenSource {
+  constructor(private tsModule: T_TypeScript) {
+    super();
+  }
+
+  get token(): VCancellationToken {
+    const token = super.token as VCancellationToken;
+    // tslint:disable-next-line variable-name
+    const Exception = this.tsModule.OperationCanceledException;
```
Comment (octref) [inline server/src/utils/cancellationToken.ts:17]: Do we have to do this assignment + TSLint rule?

File: server/src/services/vls.ts [hunk 3]
```diff
@@ -176,7 +179,7 @@ export class VLS {
     this.lspConnection.onRequest('$/getDiagnostics', params => {
       const doc = this.documentService.getDocument(params.uri);
       if (doc) {
-        return this.doValidate(doc);
+        return this.doValidate(doc).then(result => result ?? []);
```
Comment (octref) [inline server/src/services/vls.ts:182]: Make this callback async for readability

File: server/src/utils/cancellationToken.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        if (token.isCancellationRequested) {
+          throw new Exception();
+        }
+      }
+    };
+    return token;
+  }
+}
+
+export function isVCancellationTokenCancel(token?: VCancellationToken) {
```
Comment (octref) [inline server/src/utils/cancellationToken.ts:32]: Maybe call this `isVCancellationRequested` (or `isVCancellationRequestedOnToken` if you prefer)?

File: server/src/modes/script/javascript.ts [hunk 5]
```diff
@@ -107,18 +108,38 @@ export async function getJavascriptMode(
...
+        return [];
+      }
+
+      let rawScriptDiagnostics = [
+        ...program.getSyntacticDiagnostics(sourceFile, cancellationToken?.tsToken),
+        ...program.getSemanticDiagnostics(sourceFile, cancellationToken?.tsToken)
       ];
 
+      const compilerOptions = program.getCompilerOptions();
+      if (!!(compilerOptions.declaration || compilerOptions.composite)) {
```
Comment (octref) [inline server/src/modes/script/javascript.ts:136]: If the result is undefined this block won't execute, so I think it's unnecessary convert it to Boolean.

**Review by octref — CHANGES_REQUESTED** [review]

File: server/src/modes/script/javascript.ts (hunk 5 above)
Comment (yoyo930021) [reply server/src/modes/script/javascript.ts:136]: I remember it is resolve type error.

File: server/src/utils/cancellationToken.ts (hunk 4 above)
Comment (yoyo930021) [reply server/src/utils/cancellationToken.ts:32]: `isVCancellationRequested` is great.

File: server/src/utils/cancellationToken.ts (hunk 2 above)
Comment (yoyo930021) [reply server/src/utils/cancellationToken.ts:17]: This should be my mistake.
I read the tslint rules wrong.

File: server/src/services/vls.ts (hunk 3 above)
Comment (yoyo930021) [reply server/src/services/vls.ts:182]: Example add `async` in function?

**Review by octref — APPROVED** [review]

**Pull request merged**